### PR TITLE
Fix DOS_DeviceHasName function to correctly check string lengths

### LINKED ISSUE #3294: Megarace 2 Does Not Start at All and Hangs at Startup
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

Version 0.81.0-rc1

### What Operating System are you using?

Windows 10

### If Other OS, please describe

_No response_

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Windows 10, 64-Bit
https://github.com/dosbox-staging/dosbox-staging/releases/download/v0.81.0-rc1/dosbox-staging-windows-msys2-x86_64-v0.81.0-rc1.zip

In 0.81 RC, 64-bit, if you try to run **Megarace 2** at all, the game simply hangs,  goes no further, and shows no error messages at the Dosbox-Staging command prompt.  Additionally, no error messages are shown in the Dosbox Status Window.

This bug occurs with or without `-noprimaryconf` being specified. 

https://github.com/dosbox-staging/dosbox-staging/issues/2519 and https://github.com/dosbox-staging/dosbox-staging/commit/0ce4ff7e29ba53308ce231718d516dec41319ed3 were supposed to fix Megarace 2 not properly playing the game's music.  However, now the game **does not start at all** in Dosbox-Staging version 0.81.0-rc1, but simply hangs if you try to run it.

I had the same issue in 0.80.1 as user Geist22 did in https://github.com/dosbox-staging/dosbox-staging/issues/2519 where the game will run, but it crashes back to DOS prompt.

### Steps to reproduce the behaviour.

Steps to reproduce:

1. Use [Inno-Extract](https://constexpr.org/innoextract/#download) on the GOG Megarace 2 Offline Installer to extract the game folder.

2. In Dosbox-Staging, mount the newly created `app` folder produced by Inno-Extract as your C drive.

3. Next in Dosbox-Staging, imgmount `MEGARACE2.GOG` as a CD-ROM / ISO to your D drive to satisfy the CDROM check on game startup.

4. After `MEGARACE2.GOG` is successfully imgmounted as your D drive, run `MEGARACE.EXE` itself in the game folder. 

The game will simply hang, go no further, and show no error messages at the Dosbox-Staging command prompt. Additionally, no error messages are shown in the Dosbox Status Window.

### Download URL of affected game or software

https://www.gog.com/en/game/megarace_1_2

### Your configuration

_No response_

### Provide a Log

_No response_

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

The `DOS_DeviceHasName` was not correctly checking string lengths.  Root cause of the bug in #3294 was this function was matching "M" (requested name) to "MSCDEX" (device name).

## Related issues

Fixes #3294

# Manual testing

To first find this bug, I checked out commit 288dea8963221c1f9168582852c31d264185b4a2

This triggered an assertion failure on launching Megarace which led me to find the string compare bug.

I then made this change against main and confirmed the Megarace game launches without error now.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/dos/dos_devices.cpp [hunk 1]
```diff
@@ -547,17 +547,21 @@ bool DOS_DeviceHasName(const RealPt rp, const std::string_view req_name)
...
+		return false;
+	}
+
+	for (uint8_t i = 0; i < DeviceDriverInfo::name_length; ++i) {
+		const auto source = static_cast<char>(real_readb(segment, check_cast<uint16_t>(offset + i)));
+
+		// Ensure both strings have the same length.
+		// DOS name is not necessarily null terminated but padded by spaces.
+		if (i == req_length) {
+			return source <= 0x20;
```
Comment (johnnovak) [inline src/dos/dos_devices.cpp:561]: I've read this several time, and I'm still confused. I guess I could trace it with various inputs and figure it out, but this calls for clarification & simplification:

1. The comment is misleading as it implies null and spaces are used for termination. But we're checking for `<= 0x20`. So any control char is ok? Should we not be exact then and only check for 0x00 and 0x20?
2. A bit confused what happens if `req_length < DeviceDriverInfo::name_length`.
3. Can't this be moved outside of the loop? If would feel more natural and less confusing that way. `i == req_length` can only happen once the for loop is over.

Overall, I'd like to have a simpler implementation so I can see at a glance what's going on.

Comment (weirddan455) [reply src/dos/dos_devices.cpp:561]: The ‘source <= 0x20’ was taken from old code before the refactor that introduced this bug.  If you know what the DOS spec says I can change it.  Or I can try to look it up later.

To your other points, yeah I could read the DOS name into a ‘std::string’ up front and probably that would be easier to read.

Comment (johnnovak) [reply src/dos/dos_devices.cpp:561]: > If you know what the DOS spec says I can change it. Or I can try to look it up later.

I don't know, just reading this for the first time it raise a lot of questions, that's all 😄 Probably safest to adjust the comment to the code then.

Comment (weirddan455) [reply src/dos/dos_devices.cpp]: Done.  This should be easier to read if you can live with the performance penalty of copying 8 bytes into a `std::string` :wink:

Comment (kcgen) [reply src/dos/dos_devices.cpp]: Very nice fix, @weirddan455 !

**Review by kcgen — APPROVED** [review]

File: src/dos/dos_devices.cpp (hunk 1 above)
Comment (johnnovak) [reply src/dos/dos_devices.cpp]: Thanks @weirddan455. Now the intent isn't clouded by weird string micromanagement stuff. 🎉 The "performance penalty" is about on the same level as the "fuel consumption penalty" when driving the car with your hats on 😎

**Review by johnnovak — APPROVED** [review]
Nice quick fix @weirddan455 🏅 Glad to see a coding error was the culprit and not some fundamental emulation deficiency 🚀

**Pull request merged**